src/coreclr/vm/gcenv.ee.cpp

```diff
@@ -620,6 +637,7 @@ void GcScanRootsForProfilerAndETW(promote_func* fn, int condemned, int max_gen,
 #endif // FEATURE_EVENT_TRACE
         ScanStackRoots(pThread, fn, sc);
         ScanTailCallArgBufferRoots(pThread, fn, sc);
+        ScanThreadStaticRoots(pThread->GetThreadLocalDataPtr(), /*forGC*/ false, fn, sc);
```

Comment:
This is one of the perf details that I need to measure. If it's expensive, I intend to move non-collectible thread static GC base roots over to a single object array and use a normal GCHandle for that thing, and move non-collectible thread static NonGC base roots over to using malloc'd memory. The collectible ones need all this complexity, but having a pair of arrays isn't actually all that bad, and would make the higher cost here exclusive to use of thread statics on collectible types. I debated doing this on spec, but honestly, this PR is enormous enough as it is, and I wanted to stop larding on more bits and peices.. If you have concerns here about checking in without doing that work, I can add the logic to make `ScanThreadStaticRoots` really only do work in the presence of collectible assembly thread statics.

Reply:
David and I chatted briefly about the cost. it would be great if he could add a brief description of the cost model so it's easy to understand when this might start to become a perf problem and if there's any mitigation.